components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadMiddleware.kt

```diff
@@ -55,21 +55,21 @@ class DownloadMiddleware(
             is DownloadAction.RemoveAllDownloadsAction -> removeDownloads()
             is DownloadAction.UpdateDownloadAction -> updateDownload(action.download, context)
             is DownloadAction.RestoreDownloadsStateAction -> restoreDownloads(context.store)
+            is DownloadAction.AddDownloadAction -> {
+                if (!saveDownload(context.store, action.download)) {
+                    // The download was already added before, so we are ignoring this request.
+                    logger.debug("Ignored add action for ${action.download.id} " +
+                            "download already in store.downloads")
+                    return
+                }
+            }
         }
 
         next(action)
 
         when (action) {
-            is DownloadAction.AddDownloadAction -> {
-                scope.launch {
-                    downloadStorage.add(action.download)
-                    logger.debug("Added download ${action.download.fileName} to the storage")
-                }
-                sendDownloadIntent(action.download)
-            }
-            is DownloadAction.RestoreDownloadStateAction -> {
-                sendDownloadIntent(action.download)
-            }
+            is DownloadAction.AddDownloadAction -> sendDownloadIntent(action.download)
+            is DownloadAction.RestoreDownloadStateAction -> sendDownloadIntent(action.download)
```

Comment:
nit: we could combine the two branches now since they're the same `AddDownloadAction, RestoreDownloadStateAction ->`?

Reply:
The problem is that we get a compilation error, as both actions are has `action.download` but when we combine them they don't have a hierarchy relationship parent that have the `download` property.